### DISCUSSION TITLE
VB.NET support

### DISCUSSION THREAD

**Discussion by ivoryguard** [AvaloniaUI/Avalonia#10387] [Q&A] [answered]
Hello.

I have many VB.NET WPF projects and I want to reuse code snippets and WPF components in them.
I have searched vb.net support of Avalonia UI but it may be uncertain.
Can I use vb.net at Aalonia UI?
Or,  is only C# avaliable like Xamarin, ASP.NET Core, Uno and Blazor?

**Comment by Mrxx99** [comment, marked as answer]
Yes, you can use Avalonia with VB.NET.
I started porting the templates in this branch: https://github.com/Mrxx99/avalonia-dotnet-templates/tree/feature/vb
You can build this yourself to use them. Hopefully they can at one point be merged.

**Comment by DualBrain** [comment]
I noticed in a recent video by one of the devs (at a Microsoft event) that it was mentioned that VB can be leveraged with Avalonia - doing a bit of research, I stumbled upon this post. It looks like it's been almost a year since @Mrxx99 put something together to add at least some sort of initial support. @Mrxx99, have you submitted this as a pull request so that it can be considered as a contribution? If so, do you have a link to that pull request so that I can follow it?

**Reply by maxkatz6** [reply]
@DualBrain there is no "initial support". It works.
Mrxx99 was going to add default templates for VB, but it's unlikely to be accepted due to low demand, while these templates need to be maintained often.

You can create C# template and manually convert it to VB.

**Reply by DualBrain** [reply]
OK... I'll bite. So exactly how much "demand" would there need to be for this to be considered?

And on the flip side of that coin... F# has enough "demand" that it is included? Or... or... is it that someone has stepped up from the F# community to commit to maintaining the F# templates? Is that all that it would require?

**Reply by thevortexcloud** [reply]
>  So exactly how much "demand" would there need to be for this to be considered?

I can't really answer this, but I will add that the demand for VB was so low that Microsoft did not even add proper support for it in modern .NET until .NET 5. They have also somewhat openly stated they consider it to be a mostly dead language. It's mostly there for compatibility reasons with old .NET Framework codebases. 

I would also suggest if you are making a new project from scratch to move away from VB entirely. There are less people each day who know how to use VB properly.

**Reply by DualBrain** [reply]
That fact that @MikeCodesDotNET gave this response a thumbs up says volumes about the community involved with Avalonia UI. And the reality is that the entirety of the response is provably false.  Let's see... where to start...

> Microsoft did not even add proper support for it in modern .NET until .NET 5.

.NET Core 1.0 was released in June, 2016... there was unofficial support for .NET Core for VB as of July, 2016. I know this, because I was involved in this. The ability for VB to used **officially** in .NET Core happened either in v1.1 or (more probably) .NET Core 2.0; sure, it was limited in functionality... but certainly worked and was usable. The general availability of much of the "VB runtime" features officially was added in .NET Core 3.0. .NET 5 took a step forward for both VB and C# with the addition of *much of the .NET Framework* related APIs.

> They (*Microsoft*) have also somewhat openly stated they consider it to be a mostly dead language.

No. This is a gross mischaracterization of what was actually stated; twisted in a manner to, I suppose, make C# developers feel more superior to others. Not continue to evolve language is very different than not supporting it and stating that it is dead.

What was actually stated can be summed up with:

*"One of the major benefits of using Visual Basic is that the language has been stable for a very long time. The significant number of programmers using Visual Basic demonstrates that its stability and descriptive style is valued. Going forward, we do not plan to evolve Visual Basic as a language. This supports language stability and maintains compatibility between the .NET Core and .NET Framework versions of Visual Basic."*

And goes on to state:

*"Future features of .NET Core that require language changes **may** not be supported in Visual Basic. Due to differences in the platform, there will be some differences between Visual Basic on .NET Framework and .NET Core."*

Notice the word **may** in that section. Why do I draw attention to this? Because **may** doesn't say **won't** and the fact is that, sure, some new features in .NET "Core" aren't supported in VB; but there are, indeed, changes that have been made to support new features in available in "Core" that aren't in .NET Framework... and those changes are supported in VB. The above statement just makes it clear that any decision to do so will be done at the Microsoft choice and there's not going to be too much discussion about this. And even with that stated, if a friendly conversation that can clearly spell out the need... the facts are that changes have been made in this manner since this announcement. But sure... it's dead.

> It's mostly there for compatibility reasons with old .NET Framework codebases.

You are using an announcement made in a blog post back in 2020 to state that VB is only "there for compatibility reasons with old .NET Framework codebases"... where in the same blog post it was announced the including support for all of the "base" project types available. Where it goes on to state:

*"This allows Visual Basic customers to take advantage of new platform features like side-by-side deployment, cross platform support, performance and new API improvements."*

So .NET Framework is about "cross platform support"?

> I would also suggest if you are making a new project from scratch to move away from VB entirely. There are less people each day who know how to use VB properly.

Are you F'ing kidding me. Do you not realize how insulting such a comment is? It is my (as well as many others) to use VB because that is what we want to use. Who are you to dictate language choice on a platform that is, by design, about language diversity. And it continues to befuddle me how C# developers can have such a crappy attitude toward such language choice.

**Comment by MikeCodesDotNET** [comment]
@DualBrain There needs to be some clarification in how I've described our stance on VB.NET. To clarify, a more accurate term for our approach to VB.NET in Avalonia would be 'compatible' rather than 'supported'. You can definitely build Avalonia applications using VB.NET, and they will function as expected.

As for official VB.NET templates, we are not planning to introduce them. This decision is influenced by several factors, including the tiny user base of VB.NET and the resources required for the continued maintenance of such templates.

Our support for F# templates is driven by our evaluation of the language's trajectory and future potential. Unfortunately, the outlook for VB.NET does not suggest it would be a prudent investment for us.

In light of this, I agree with @thevortexcloud's recommendation to consider transitioning away from VB.NET. It's a suggestion borne out of practicality and an eye towards future-proofing your development work.

**Reply by DualBrain** [reply]
I appreciate the clarification of "compatible" versus "supported". However, I also feel that it is short-sighted in it actively demonstrates that although there is clearly interest in Avalonia UI by VB developers, those developers are *effectively* shunned and project members demonstrate that we are not welcomed. I really don't understand why an open-source project would actively reject the opportunity to collaborate with a segment of the .NET community that is clearly interested. But by these sorts of responses, the lack of project templates included (regardless of how that happens) and the active rejection of any sort of contributions certainly seems to *say something*. And all of this started off with the statement "unlikely to be accepted due to low demand". With the attitude that seems to be pervasively shared by the Avalonia UI maintainers; is there any wonder why there is "low demand"?

And this is such a shame. Watching the video showing the approach... I really do feel like Avalonia UI has something good going for it. I truly feel like the overall design and approach to the cross-plat problem is far better than any of the other projects I've seen. It's just really sad that *we* aren't welcome. And if we do want to do it... the response is basically (and I'll paraphrase):

> Learn C# and shut up!

So I'm not sure what I want to do at this point. Do I chuck Avalonia UI in the garbage knowing full well that, although it's a great platform, that any contributions that I may make are going to be met with animosity, criticism and constant "learn C# and shut up" sort of responses. Why would I do that? Furthermore, why would I encourage others to want to do the same?

Bottom line, I think you are wrong in your assessment that there is "low demand". There is a huge difference between "disinterest" and "simply not knowing it is an option". If VB developers aren't aware that it is even a possibility; how are they supposed to get interested? And when they do get interested, they are demeaned and reminded out "insignificant" they are.

Yup... continues to wonder why there is "low demand".

Like I said, I'm having a hard time trying to decide what further investment I'll do with Avalonia UI... it does look like a wonderful platform - it's just unfortunate the general attitude I'm getting.

(And, to think, all that was being asked was to not have to manually convert C# templates to VB - instead, our community contribute the necessary VB templates... and even offer to maintain them.)

**Reply by TheWitheredStriker** [reply]
Mike, could it be that perhaps the reason for a lack of VB users trying to use Avalonia is the fact that not only Avalonia doesn't work well with VB.NET, but also the fact that its maintainers appear to be openly opposed to addressing it? Might the disinterest of VB.NET users in Avalonia be a circular network problem?

I love C#, F# _and_ VB.NET. .NET, as its core, is ment to be a software platform that supports all three languages equally. I love F# just as much as the others, but if you will accept F# templates despite the fact that C# is far ahead in popularity and commerical use, then I believe there is no issue with accepting VB.NET templates either, especially not when someone else has gladly done all the work for you out of raw passion and is quite literally just offering them to the project, requiring only green light. As DualBrain has stated, we will also gladly assume responsibility for their maintenance, and are willing to even do so all on our own if so be it.

I apologize if I come across as rude or ignorant, but I politely urge you to reconsider.

**Reply by MikeCodesDotNET** [reply]
@DualBrain, I appreciate your feedback, but let's address the facts. The usage of VB.NET within the developer community has significantly dwindled. This is not an opinion but a reality reflected in metrics such as the [2023 Stack Overflow Developer Survey](https://survey.stackoverflow.co/2023/#technology), where VB.NET is notably absent. The demand for VB.NET support with Avalonia UI is minimal, and this isn't conjecture; it's based on the actual engagement and requests we receive.

While we respect your preference for VB.NET, accusing us of alienating VB.NET developers is unfair. We simply prioritise our resources in higher-demand areas. Community contributions are always welcome, but they also require our oversight, adding to our workload. For this reason, we've suggested you create and maintain the templates, and we'll include links in our docs. 

We hope you continue to see the value in what Avalonia UI offers, even if it doesn't align perfectly with your preferences.

**Reply by DualBrain** [reply]
Telling me to convert to C# is also "unfair". I totally respect that you don't want to add the templates. Fine. I appreciate the fact that you will provide a link to these templates if they exist elsewhere. That is wonderful. But the simple fact is that I'm being asked to abandon my language of choice due to what amounts to a serious misunderstanding of the realities of a community that I'm actually a part of.  To point to SO as the answer... that is beyond ludicrous; let's see let's ask a hostile against VB what they think of VB... uhmmmmm... I wonder how that would fair.

And, again, the requests you receive... if the responses are "you should use C#"... how many more of these requests do you think you are going to get?

There are many of us here interested. There several of us willing to spend our time contributing. If you don't want this contributed to the project directly, that is fine. And I can certainly respect that. I also appreciate (again) that providing a link to where we can do this elsewhere... if just for the discoverability of having such being increased. 

But I certainly 

> I would also suggest if you are making a new project from scratch to move away from VB entirely. There are less people each day who know how to use VB properly.

> In light of this, I agree with @thevortexcloud's recommendation to consider transitioning away from VB.NET. It's a suggestion borne out of practicality and an eye towards future-proofing your development work.

These *suggestions* and *recommendations* are not only unwarranted, unwanted and unnecessary... they are certainly do not come across as "respectful". Clearly there are several of us here that are passionate about using VB... even if you think our numbers are small. Please don't dismiss or demean us... and certainly, please don't tell us that we should abandon *our* language simply because *you* think it isn't worthwhile.

I'm certain the VB community can move forward with creating and hosting a nuget package... this is wonderful. But I also feel like there needs to be conversation about respecting that we do indeed exist and should not be "educated" on the perceived state of our community when clearly none of you are interested in enlightening yourself with regards to what is going on in our community.

As I said, really like what is being done in Avalon UI and am excited to help where I can to share/educate/encourage its usage in the VB community; but I certainly feel like the negativity toward VB should be nipped-in-the-bud. Why would I in good conscience encourage others to engage if they will be met with the superiority complex exhibited?

I suppose all I'd like to get at this point out of this thread is the following:

- agree that it will not be included in the distribution...
- however, you can point to a nuget package that adds these (hosted elsewhere, maintained elsewhere)...
- a commitment going forward that simply because of our choice of language we will not be criticized.

It is this last part that I'm most annoyed by. We, the VB community, certainly have no right to tell any open-source project what to do. Can we ask? Sure. Can we request clarification? One would hope so. Can we come to some common ground? One would think so. But for it to end up with essentially where this ended up... that's not acceptable.

If this thread (as well as the other) had been limited to the first two items... all would be awesome! Clarification and clear direction on how the VB community should move forward. No hard feelings and we can easily progress. But the stabs to the gut were unwarranted; again, it's the CLR/CLS/CLI ".NET"... multiple languages regardless of your personal (potentially misinformed) opinion of these other languages should not result in negative responses. We do exist... we are interested... even if that isn't enough to warrant your consideration... doesn't warrant us being told we should abandon and switch to what your choice was.

In the end, just a reminder... contrary to popular opinion "CLR" does not stand for "C# Language Runtime".

**Comment by RealDoigt** [comment]
Hello,
at my work we use VB .Net, please don't be stubborn and accept this free help which will make my life easier.
